### ISSUE TITLE
[Evergreen] Improve SQL error messages

### ISSUE THREAD

**Issue by chaas** [MaterializeInc/materialize#15594]
Short Term Objectives
[] Formalize the SQL error standard (based on https://www.postgresql.org/docs/current/error-style-guide.html). This involves documenting the guidance in our developer docs and getting agreement across engineering, devex, and support.
[] Apply the standard to existing issue #15502 

We should use this ticket to track additional errors that should be improved according to the standard.

Longer Term Objectives
[] Broad refactor of error handling to get rid of `Unstructured` SQL errors and the use of `anyhow::Error` in any code that's part of the core database engine (see larger discussion here https://github.com/MaterializeInc/materialize/discussions/5340)

**Comment by chaas** [comment]
@benesch resuscitating your old discussion on SQL errors and turning it into an actionable epic/evergreen task.
I am planning to move forward now with declaring the postgres standard as our own and applying it to the unknown type error.

I'd like your take on a broader refactor.

cc @ggnall

**Comment by benesch** [comment]
Thanks for digging that up, Cara! I totally forgot I’d written that up years ago. 

The good news is the big refactor of the SQL and coordinator layers came to pass! The structured enum errors _do_ exist. What’s remaining is a very, very long road to stamping out every usage of `anyhow::Error` and replacing it with a properly structured error. IMO what remains is an essentially an evergreen project to improve each error message one by one. I’m very supportive of a big push to fix the next ten, twenty, fifty, whatever of these error messages, though I think each of the fixes can be a small focused PR. Is that what you had in mind?

**Comment by chaas** [comment]
Yeah we still default to using the `!sql_err` macro in a lot of places, which creates a `PlanError::Unstructured`, which is technically not a `anyhow::Error`, but still is essentially just a single error string. Agreed it's a long road to move them all over one by one, and I'm not sure of any more systematic way.

I've added a section to the epic/evergreen description "Ticket Tracker for Individual Errors to Improve" where folks can call out individual errors as they come across them.
